### DISCUSSION TITLE
Is it possible to resize a client from the compositor?

### DISCUSSION THREAD

**Discussion by xTrayambak** [CuarzoSoftware/Louvre#90] [Support] [answered]
Hello. I've checked the docs for LSurface and couldn't find any function to set a surface's size, only its position. Is there a way to force a surface to resize on the compositor side?

Sorry if this is an obvious question. I'm just not very well versed with the Louvre API.

**Comment by ehopperdietzel** [comment, marked as answer]
Hi! I noticed you started developing a compositor some time ago and now you're creating bindings for Nim. That's very cool!

A surface on its own only contains a buffer (texture) and some properties like scaling factor, transform, etc. What gives a surface additional "functionality" is a role, which is assigned by the client. Without a role, surfaces shouldn't be displayed (i.e., [LSurface::mapped()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_surface.html#ac563c8ace44bf8111c0ee1c034e967e2) returns false). When a client assigns a role to a surface, the [LSurface::roleChanged()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_surface.html#ad277475bd3c8656d83e5d221e5db4bf0) event is triggered. You can access the role with [LSurface::role()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_surface.html#adcb0a0c7570cedec2cc38822943a7fff) or with role-specific functions. For example, [LSurface::toplevel()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_surface.html#a744809727c631248ae26aae91fa08a6d) returns a pointer to an `LToplevelRole`, or `nullptr` if the surface has another role.

There are many different roles which are listed [here](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_surface.html#details). The one you're probably looking for is [LToplevelRole](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#details), which represents typical desktop windows. The `LToplevelRole::configureXXX` methods allow you to set the size of a window, change its state (maximized, fullscreen, etc.) and more. This is an asynchronous operation, to know when the client has applied your configuration, you should override the [LToplevelRole::atomsChanged()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#a8ddd427bdb9cacae8f788b3b11554f9a) event, which lets you handle many of its properties' changes atomically.

For example:

```cpp
// Somewhere in your code

#include <LSurface.h>
#include <LToplevelRole.h>

if (surface->toplevel()) 
    surface->toplevel()->configureSize(800, 600); // Not applied immediately
```

```cpp
// In your own toplevel implementation
void YourToplevelRoleImp::atomsChanged(LBitset<AtomChanges> changes, const Atoms &prevAtoms) {
    if (changes.check(WindowGeometryChanged)) {
        // Now windowGeometry().size() should be (800x600) (if the client obeyed your request)

        // And what was the previous size? : prevAtoms.windowGeometry.size()
    }
}
```

I understand that the documentation can be overwhelming and difficult to navigate. As a tip, I recommend always clicking on the `More...` link at the top of each class page. That will take you to a brief description of the core concepts of each class, making it easier to understand.

**Comment by xTrayambak** [comment]
Thank you for the quick answer! Louvre is quite the joy to write code with, I must admit. I've used wlroots before, and Louvre is leaps more friendlier than it. :^)

**Reply by ehopperdietzel** [reply]
I am glad you enjoy using it! 😄
